lib/geordi/settings.rb

```diff
@@ -67,47 +67,34 @@ def read_settings
       local_path = LOCAL_SETTINGS_FILE_NAME
       firefox_path = File.expand_path("../../.firefox-version")
 
-      if File.exists?(global_path)
-        global_settings = YAML.safe_load(File.read(global_path)) || {}
-        check_for_invalid_keys(global_settings, ALLOWED_GLOBAL_SETTINGS, global_path)
-      end
-
-      if File.exists?(local_path)
-        local_settings = YAML.safe_load(File.read(local_path)) || {}
-        check_for_invalid_keys(local_settings, ALLOWED_LOCAL_SETTINGS, local_path)
-      end
-
-      if File.exists?(firefox_path)
-        Geordi::Interaction.warn "Unsupported setting \".firefox-version\". Please remove the file in #{firefox_path}."
-        exit 1
-      end
+      global_settings, local_settings = validate_configuration(global_path, local_path, firefox_path)
 
       @global_settings = global_settings || {}
       @local_settings = local_settings || {}
     end
 
+    def validate_configuration(global_path, local_path, firefox_path)
+      return if ENV['GEORDI_INVALID_SETTINGS_WARNED']
+
+      global_settings = File.exists?(global_path) ? YAML.safe_load(File.read(global_path)) || {} : {}
+      check_for_invalid_keys(global_settings, ALLOWED_GLOBAL_SETTINGS, global_path)
+
+      local_settings = File.exists?(local_path) ? YAML.safe_load(File.read(local_path)) || {} : {}
+      check_for_invalid_keys(local_settings, ALLOWED_LOCAL_SETTINGS, local_path)
+
+      Interaction.warn "Unsupported config file \".firefox-version\". Please remove #{firefox_path}." if File.exists?(firefox_path)
+
+      ENV['GEORDI_INVALID_SETTINGS_WARNED'] = 'true'
+      [global_settings, local_settings]
+    end
```

Comment:
Thanks for refactoring this. However, I apologize: I did overlook that loading the settings was done in L71 and L76. Moving the loading to `validate_configuration` does not make much sense.

So, lets elaborate on your last suggestion:

```ruby
SETTINGS_WARNED = 'GEORDI_INVALID_SETTINGS_WARNED'
# ...

def read_settings
  global_path = GLOBAL_SETTINGS_FILE_NAME
  local_path = LOCAL_SETTINGS_FILE_NAME

  global_settings = if File.exists?(global_path)
    YAML.safe_load File.read(global_path)
  end

  local_settings = if File.exists?(local_path)
    YAML.safe_load File.read(local_path)
  end

  # Prevent duplicate warnings caused by another instance of Settings
  unless ENV[SETTINGS_WARNED]
    check_for_invalid_keys(global_settings, ALLOWED_GLOBAL_SETTINGS, global_path)
    check_for_invalid_keys(local_settings, ALLOWED_LOCAL_SETTINGS, local_path)
  
    firefox_path = File.expand_path("../../.firefox-version")
    if File.exists?(firefox_path)
      Interaction.warn "Unsupported config file \".firefox-version\". Please remove #{firefox_path}."
    end

    ENV[SETTINGS_WARNED] = 'true'
  end

  @global_settings = global_settings || {}
  @local_settings = local_settings || {}
end
```

`check_for_invalid_keys` then needs to return if `settings` is `nil`.

Please extend your Cucumber scenario for the `.firefox-version` warning. I doubt the code is correct, since Geordi is mostly executed in project root. `File.expand_path('../../.firefox-version')` would not find that file in project root.